### DISCUSSION TITLE
asyncio on a second thread

### DISCUSSION THREAD

**Discussion by bjpirt** [micropython/micropython#12423] [Using MicroPython] [answered]
Hi all,
I'm currently trying to run a second thread on an ESP32 running asyncio (but with no asyncio on the main thread currently). I've seen references on the internet about how asyncio doesn't work on threads - before I start debugging why my code isn't working I just wanted to check if there are still issues with this arrangement?
Thanks,
Ben

**Comment by andrewleech** [comment]
Currently in micropython there's only one global asyncio event loop, I don't think it will be possible to have a separate loop on a second thread.

**Reply by jimmo** [reply, marked as answer]
@bjpirt As long as you only run a single event loop, there's no problem. (@andrewleech -- asyncio doesn't know about threads and doesn't have any thread-local storage, it doesn't care _which_ thread it's running on).

That said, you need to be very careful that you do not access/modify any asyncio internal state concurrently (e.g. from another thread or interrupt handler). e.g. you may not call `create_task` from other threads, or `Event.set`, while asyncio is running (on any thread). MicroPython has asyncio.ThreadSafeFlag which you can use to safely signal into asyncio from another thread or interrupt handler.

**Reply by andrewleech** [reply]
Thanks @jimmo for the extra detail. I assumed the question was about being able to run two separate loops concurrently in separate threads; this is possible on cpython but not on micropython.

The single event loop will only be serviced / async functions run on the thread it was started on.

**Comment by peterhinch** [comment]
@bjpirt [This discussion](https://github.com/orgs/micropython/discussions/11999) was about running `asyncio` on core 1 and delegating work to core 0. While clever, I'm unconvinced that it is an effective way to design applications.

There can only be one instance of `asyncio`. Core 0 has privileged access to hardware. ISR's run on core 0. In my opinion the best design is to run `asyncio` on core 0, delegating activities to core 1 only where they would otherwise stall core 0. This takes advantage of the excellent capability of `asyncio` to handle concurrency in hardware interfaces while using core 1 to do those things where it would fail, namely dealing with blocking methods and demanding computational loads.

**Reply by bjpirt** [reply]
Thanks @peterhinch - that's good info about the Core 0 priority

**Comment by bjpirt** [comment]
Thanks for the replies all - so I should be good to run a single instance of uasyncio on Core 1, but it's better to use Core 0 if I need to access hardware.

**Comment by andrewleech** [comment]
There's also some previous discussion about separate thread loops here: https://github.com/micropython/micropython/issues/8340
